### PR TITLE
Added prefix and suffix to columns

### PR DESCRIPTION
**Pull request by promatik**
Added prefix and suffix on columns I thought it made sense.

Columns already had prefix and suffix;
- [x] array_count
- [x] model_function_attribute
- [x] model_function
- [x] relationship_count
- [x] row_number
- [x] text

Columns updated;

- [x] array
- [x] boolean
- [x] check
- [x] closure
- [x] custom_html
- [x] date
- [x] datetime
- [x] email
- [ ] ~image~
- [x] json
- [x] markdown
- [x] markdown
- [x] multidimensional_array
- [x] number
- [x] phone
- [x] radio
- [x] select
- [x] select_from_array
- [x] select_multiple
- [ ] table (?)
- [x] textarea
- [ ] ~upload_multiple~
- [ ] video (?)
- [ ] view (?)

On `image` and `upload_multiple` prefix is used for URL prefixing, I think it didn't make sense to use prefix and suffix anyway ...
On `table`, `video` and `view` I thought it doesn't make sense to use prefix and suffix.

I'm up to your thoughts on this, we may revert some columns or add other 🙌

### CODE REVIEW

**Review by tabacitu — COMMENTED** [review]
Very clean - nice job @promatik , I love it when the changes are surgical 👍 I'll let @pxpm do the final test and review, I have only one minor concern from looking over the code:

**What should we do about columns that contain multiple values?** Columns like `array`, `multidimensional_array`, `select`, `select_from_array`, `select_multiple` - those that have `foreach` and spit out multiple values? **Should the prefix be _general_ (for the entire table cell), or individual for each value?** What would be most intuitive? What would be most useful?

Rephrased: should devs be able to use:
- prefix `Items` to get `Items One, Two, Three`
- (or) prefix `Item` to get `Item One, Item Two, Item Three`

I think I agree with the first option (the way it's already implemented now by @promatik - so no changes needed ). If we keep it this way, the dev could theoretically achieve the second one too, by using the `select_from_array` field. So they kind of get both options if they really want per-entry-prefix and per-entry-suffix. Just wanted to express this minor concern so that we have it in mind before merging, and get your thoughts on it.

**Comment by promatik** [comment]
I actually started by adding the prefix and suffix to each entry on lists, but after the first test it made more sense to do it "globally", **one prefix and one suffix per column**, it's more standard across all columns.

The only thing that turns this "less standard", is that every prefix and suffix are next to the content, inside the wrapper.
And in these cases (lists) it's outside the wrapper.

Anyway, since it's nonsense to add one wrapper for the prefix and another for the suffix, I approve it this way 😅

**Comment by pxpm** [comment]
@promatik @tabacitu 

My two cents here, is that the `prefix` and `sufix` apart from beeing used to configure the display string `Items 1, 2, 3` it is also used to configure the display property of each item like you already said.

In your example it was `item 1, item 2 ...` but what I think suffix/preffix really shines is when you need to have a special display like:

`KEWK_value, KEWK_value2` or `#1, #2, #3`  **(I use this last one alot of times, usually setting up a model accessor)**

Maybe it's time to do what we did to wrapper ? Move columns suff/pref to a separate blade file and do the logic there. 

Maybe we could introduce a `column_suffix_per_entry => true` and `column_prefix_per_entry => true` 

Also maybe make it `callable`. 

I think doing this we are guarantee we are not going back look at this for a lot of time, leaving it as flexible as possible. 

I think one major benefit of this would be reducing the number os `accessors` we have in our models just for example to add a `+` (plus) or `-` (minus) into the string we want to display. 

This stops making much sense if you think about `text` column for example, that you can replace with a `closure` one with relative ease, but still.

Let me know what you guys think.

**Comment by promatik** [comment]
@pxpm now that I think once more about it, maybe prefix and suffix should work as the wrapper, per entry when there's a list.
This is a huge dilema 🤔
When I did it I thought about doing something like `"Selected words are A, B, C."`, and I thought about consistency on the behavior of all the columns.

> Maybe we could introduce a column_suffix_per_entry => true and column_prefix_per_entry => true

I'm afraid of introducing too much complexity, but I mostly agree with turning this as much configurable as possible 🤷‍♂️
Maybe `prefix` and `prefix_entry` on lists, and something similar to wrapper.

Btw, columns need a decent refactor on v4.2, there are other things like text limit that should be more consistent.

---
Final thoughts on this PR; Aprove it as is from 4.1, and break everything on 4.2 😅 with `prefix`, `prefix_entry`, `suffix`, `suffix_entry`, `wrapper`, `wrapper_entry`, all can be set as string or callable.

Let's hear @tabacitu ✌

**Comment by tabacitu** [comment]
Hmmm.... I think `column_suffix_per_entry` and `column_prefix_per_entry` is overkill (I agree with @promatik that it adds too much complexity). But I also agree with @pxpm that it's something you sometimes need, and we should be as flexible as possible. So yeah... ideally we should empower the developer to do both: 
- add stuff before ALL entries
- add stuff before EACH entry

At first I thought they could already do `before EACH entry` using `wrapper` or `value` but after a second look... I don't think that's the case. Please correct me if I'm wrong - if some of our existing features already allow you to easily do that.

Assuming I'm right, how could devs do both of that, ideally?
- they could use `prefix` and `suffix` to add stuff before ALL entries (I think this is the intuitive thing to do, and it's already implemented; we just merge this as-is if I understand correctly and it's already well-tested; plus there's no better way to add this functionality than with the already-existing `prefix` and `suffix` which work exactly this way for single-item columns)
- since we already have an attribute that modifies each entry (`wrapper`), what if we add `prefix` and `suffix` to the  existing `wrapper` attribute too, so that it adds stuff to each entry?

The when with the following code:
```php
$this->crud->addColumn([
  'name'      => 'items',
  'prefix' => 'Items ', // add stuff before ALL entries
  'suffix' => '.', // add stuff after ALL entries
  'wrapper'   => [
      'element' => 'a',
      'prefix' => '[', // add stuff before EACH entry
      'suffix' => ']', // add stuff after EACH entry
      'href' => function ($crud, $column, $entry, $related_key) {
          return backpack_url('item/'.$related_key.'/show');
      },
      'target' => '_blank',
      'class' => 'badge badge-info',
  ],
]);
```
You'd be getting:
`Items [1], [2], [3].`

So... the best of all world, without adding the ugly `column_suffix_per_entry`, right? What do you guys think? If you like this, I say let's merge the _before ALL entries_ functionality right now, and create a different issue for 4.2 to remember to continue this work and add the _before EACH entry_ functionality.

---

Just realised this could potentially open up another question. Is the wrapper prefix BEFORE the element or AFTER the element? So if it's a link are you getting:
- [[1](http://google.com)], [[2](http://google.com)], [[3](http://google.com)]
- (or)
- [[1]](http://google.com), [[2]](http://google.com), [[3]](http://google.com)

Regarding this matter, I think it makes sense to do the `wrapper[prefix]` and `wrapper[suffix]` both INSIDE the element. I think that's what most people need, most of the times. _If_ we think outside is also needed, we could _maybe someday_ add `wrapper[before]` and `wrapper[after]` for that. But... somehow adding these from the start seem overkill to me, I don't know why...

**Comment by promatik** [comment]
I agree 🙌
I haven't came up with that solution before `wrapper[prefix]` and `wrapper[suffix]`, but I think it is the best option.

And yes, inside the element makes sense
> Items [[1]](http://google.com), [[2]](http://google.com), [[3]](http://google.com).
> The list is [#1](http://google.com), [#2](http://google.com), [#3](http://google.com).

---

Something we could also provide **(in v4.2)**, is the `wrapper[separator]`, for lists, `, ` by default.

With that, if you really want something like;
> The list is ([#1](http://google.com)), ([#2](http://google.com)), ([3](http://google.com)).

_Note the `(` `)` are outside the wrapper, `#` is inside._

Developers could do;

```php
$this->crud->addColumn([
  'prefix' => 'The list is (',
  'suffix' => ').',
  'wrapper'   => [
      'prefix' => '#',
      'separator' => '), (',
```

Kind of tricky, it wouldn't even be document, but developers would realize they could do it easily if they need to.

**Comment by promatik** [comment]
@tabacitu as you previously said, and I agree, this PR is good as is. So it allow us to add prefix and suffix for the whole column.

I created this branch;
https://github.com/Laravel-Backpack/CRUD/compare/columns-prefix-and-suffix-on-wrapper for the prefix and suffix on wrapper.

I tested and it's working as expected, in my opinion we could merge both 🤷‍♂️

I'll let you decide whether to merge in this, or create a new PR.

**Comment by tabacitu** [comment]
Excellent idea with `wrapper[separator]`, @promatik . Let's do a separate PR and assign it to 4.2 please. It's a feature. Plus there are some thing to talk about there too, like whether we need `spans` or not, thing like that. Problem for future us.

---

Note that this isn't "_ready to merge_" just yet, it doesn't have Pedro's approval yet. Caught 2-3 issues this week.

File: src/resources/views/crud/columns/array_count.blade.php
Comment (pxpm) [inline src/resources/views/crud/columns/array_count.blade.php:16]: I think for consistency we should not make the `suffix` here have a space by default from the column value.

File: src/resources/views/crud/columns/array_count.blade.php
Comment (pxpm) [inline src/resources/views/crud/columns/array_count.blade.php:8]: For future us, we should remove this hardcoded value `-` (dash)!

File: src/resources/views/crud/columns/array_count.blade.php
Comment (pxpm) [inline src/resources/views/crud/columns/array_count.blade.php:7]:
```suggestion
    $column['suffix'] = $column['suffix'] ?? ' items';
```

**Review by pxpm — CHANGES_REQUESTED** [review]

File: src/resources/views/crud/columns/array_count.blade.php
Comment (promatik) [reply src/resources/views/crud/columns/array_count.blade.php:7]: @pxpm this file is problematic.
By default it had the space before the the word `items`. So we decided to keep it in v4.1.
The worst part is the space was hard-coded not in the suffix, but in the code.

So there is a patch already for 4.2;
https://github.com/Laravel-Backpack/CRUD/pull/3436/files

File: src/resources/views/crud/columns/array_count.blade.php
Comment (promatik) [reply src/resources/views/crud/columns/array_count.blade.php:16]: Please check above why this space is here 😅

File: src/resources/views/crud/columns/array_count.blade.php
Comment (promatik) [reply src/resources/views/crud/columns/array_count.blade.php:8]: I agree 🙌
That's why I moved all this default values (the dashes) to the top 😅
Maybe this should be set on configs 🤷‍♂️

We should do it when we refactor fields/columns on v4.2.
There are some things to improve on this matter.

File: src/resources/views/crud/columns/array_count.blade.php
Comment (pxpm) [reply src/resources/views/crud/columns/array_count.blade.php:7]: Thanks for clarification. :)

**Review by pxpm — APPROVED** [review]

**Comment by tabacitu** [comment]
Excellent work here guys! 👏👏👏
Merged!

**Pull request merged**
